Small redesign of style manager table

abaevbog opened the pull request:
- Remove the delete button below the style table
- Make selected rows stand out less. They will only have an outline, like focused rows. This should remove the impression that this is the table where one selects the style for citations.
- Add minus icon buttons to each row to delete the style without having to select the row.

Addresses: https://github.com/zotero/zotero/issues/5511

dstillman commented:
1. Is there any reason the rows need a focus ring at all when clicked? I get that we need to announce focused rows for screen readers, but what purpose does showing the focus have otherwise? It's the pivot row for subsequent keyboard operations, but I'm not sure that needs to be conveyed. This is better, but I feel like people could still think they've selected a style this way. Can we show a focus ring only for keyboard navigation?
2. There doesn't seem to be a way to remove rows via the keyboard.

abaevbog commented:
> Is there any reason the rows need a focus ring at all when clicked? I get that we need to announce focused rows for screen readers, but what purpose does showing the focus have otherwise?

I assume one could click into the table and then want to use arrowUp/Down to go through the list of rows - especially if the list is long. In that case, I feel like we do need to indicate which row you are starting at for this navigation to make sense? Also, it is possible to select multiple rows in the table to delete them in bulk, so we need to tell which rows are selected for that too.

https://github.com/user-attachments/assets/ca43641d-7bd6-4c50-9c0c-6c1c02ebdd9e

I suppose we can try to drop multi-selection here and show the outline only if keyboard navigation happens? Then, nothing would appear on click of a row, which is what most people would do.

> There doesn't seem to be a way to remove rows via the keyboard.

When a row is selected, does Backspace not do it on your end?

dstillman commented:
> I assume one could click into the table and then want to use arrowUp/Down to go through the list of rows - especially if the list is long.

Yeah, that's what I meant about the pivot row, but I'm just not sure it justifies the focus ring, at least given the pervasive confusion this has caused.

I didn't think of multi-selection and Backspace, though. I do think we could drop that, particularly since it doesn't really make sense visually anymore. Now that there's a per-row button, I was expecting to be able to tab to that, and it didn't occur to me that I could press Backspace. E.g., we don't support Backspace in the tabs menu, which has a per-row close button — you have to tab to each row's close button.

I'm sure there are people who delete all but the one citation style they ever use, and I guess they'd be annoyed by having to click and confirm on each row, but we've never really catered to that — I believe we automatically restore all bundled styles on every app update. (That's arguably not ideal, but I think it's what's always happened.) We could also do the same optimization we have in the tabs menu where focus moves to the close button of the next row after a delete.

(We should perhaps also move "Reset Styles…" out of Advanced and into to this section to make it easier to get back to the default set.)

abaevbog commented:
> I didn't think of multi-selection and Backspace, though. I do think we could drop that, particularly since it doesn't really make sense visually anymore.

Cool, so if we are willing to drop it, we can get the following interface. Clicking on anything in the table (that's not the button) will do nothing and no selection/outline appears. Clicking on the icon button will still show the confirmation dialog. Finally, if you Tab into the table (as determined by `:focus-visible`), then the outline will appear and you can use arrows to move it. Clicking on the table will still have no effect there. I assume the majority of people would never end up tabbing into the table, so most people would never see the outline.

https://github.com/user-attachments/assets/5bf104e8-e7b0-4b14-81f2-82e3b06c70a7

>  Now that there's a per-row button, I was expecting to be able to tab to that, and it didn't occur to me that I could press Backspace. E.g., we don't support Backspace in the tabs menu, which has a per-row close button — you have to tab to each row's close button

Backspace to remove the style is the original behavior from before this change, so I thought there's no need to change it. Also, we do have other instances of Backspace/Delete on table rows performing some kind of "Remove"-like action: in library tab, Backspace on a row removes items from collections (or trashed them), in citation dialog, backspace on a row will remove all bubbles of that item from the citation, and in the (in-works) redesigned edit bibliography dialog, backspace will remove the item from the bibliography. Not to mention that pressing Backspace vs Tab (to focus the `-` icon) + Space/Enter (to click it) is just a bit faster.

Finally, the virtualized table itself receives the focus (not the rows or cells), so focusing the `-` button would go against that existing setup. We can totally rework it though so that Tab focuses the `-` icon of the selected row if it's worth it? But in that case, that would need to apply to other tables with icons, right? What would happen with tables where multi-select is enabled (e.g. citation dialog with `+` button)?

> I'm sure there are people who delete all but the one citation style they ever use, and I guess they'd be annoyed by having to click and confirm on each row

> We could also do the same optimization we have in the tabs menu where focus moves to the close button of the next row after a delete.

If one wanted to delete many styles with this change, one could tab into the table and then repeatedly press Backspace+Enter. It's not the most effective way to bulk-delete anything but faster than clicking the `-` icon on each row. Virtualized table already preserves the selection, so we end up at the next/previous row when something is deleted.


https://github.com/user-attachments/assets/39a0e41e-7665-42ee-8c86-ba09aa9e1499

If we wanted to make it possible to delete rows faster, we could skip the confirmation dialog if Command/Ctrl is being held?

abaevbog commented:
(As soon as we are set on the desired behavior here, we should make sure to run this by adomasven as well - https://github.com/zotero/zotero/pull/5518/commits/36acf2f91f2e9e649d270a56a87d0a76e2da9dac removes the earlier pointer-events workaround from the `virtualized-table`. As far as I can see, everything works as expected now without it but it's best to double-check).

abaevbog commented:
> Not being able to select styles certainly help, but adding some text description below Style Manager is probably the most straightforward way to clear things up. Something like "Styles in this list will be available for quick copy, creating bibliography, and word processor plugins"?

Good point! The exact phrasing is tricky because we essentially need to say that this is not where you select the style. But it's hard without saying where you DO select the style in that case. But that is different depending on the feature you are using (e.g. quick copy vs document preferences). "Styles in this list will be available for quick copy, creating bibliography, and word processor plugins" also doesn't necessarily tells me that I don't get to select the style right here.

So I wanted to suggest the following phrasing: "This table lists your available styles: you can add new styles or remove styles you are not using. Whenever you need to select a style to cite with, these will be presented as available options." I feel like this says what you CAN do in the table, and then states that these will be available options when you are ASKED to select one. How does that sound? It's very much possible that I have really overthought this and "Styles in this list will be available for quick copy, creating bibliography, and word processor plugins" is more concise and just better.

> I think it will help if we reword the latter to "Add from Zotero Style Repository...", make it into a button instead of a link, and put both Add buttons on the bottom left corner of the table.

I fully agree. Done.

> Also, let's add zebra stripes to the table, so that it's easier to tell which style you are removing.

Done.

This is what it looks like at the moment. (Again, the description phrasing can be tweaked more).

<img width="591" height="410" alt="Screenshot 2025-08-26 at 4 02 19 PM" src="https://github.com/user-attachments/assets/8e319dbf-c190-421e-a766-5d067b4cf5d6" />

dstillman commented:
I'll tweak the description text, though I doubt it will help much. People don't read things.

For the buttons:

1. "Add from Style Repository…" is too technical. I think we should stick to "Get Additional Styles…", even if it's inconsistent with "Add from File…".
2. "Add from File…" is much less important. The vast majority of people will be adding from the styles page. So while I agree that the previous link approach was maybe not prominent enough, I don't know that we want to fix that by making the options equally prominent. Maybe @yexingsha has some other idea here for how to differentiate the two stylistically.
3. The horizontal order should be reversed, with "Get Additional Styles…" on the right. That's where the default button goes on macOS. If we really care, we could swap the order of the two buttons on the right on Windows, with "Get Additional Styles…" to the left of "Add from File…".
4. Maybe "Reset Styles" → "Restore Default Styles…"?

yexingsha commented:
<img width="1000" height="800" alt="settings" src="https://github.com/user-attachments/assets/dc2cf316-3c1d-4149-ad2f-e986617f6672" />

Do you think a split button will be too confusing? The idea is that if "Add from Repository" is selected (by default) the button will say "Get Additional Styles...", otherwise it will say "Add from File...".

dstillman commented:
1. As a button, it needs to be title cased — "Get Additional Styles…".
2. "Add from File…" should be grouped with "Get Additional Styles…" on the right, not "Restore Default Styles…".
3. "Get Additional Styles…" is always in the right-aligned group, regardless of platform. On Windows, it should just be to the left of "Add from File…". I'm not sure whether "Restore Default Styles…" should still be on its own on the left on Windows or if it should be the right-most option after "Add from File…" — we can try to find an equivalent example from Microsoft. (I noticed that a right-aligned `[OK] [Cancel] [Apply]` in one dialog, but I'm not sure if there are any where an auxiliary option is on its own on the left.)

abaevbog commented:
> As a button, it needs to be title cased: "Get Additional Styles…"

Ah, sorry about that. I didn't realize the old `.dtd` string didn't have proper capitalization. Should be fixed now.

> "Add from File…" should be grouped with "Get Additional Styles…" on the right, not with "Restore Default Styles…" on the left.

I see, I moved those two buttons together. This is what it looks like on macOS:

<img width="595" height="368" alt="Screenshot 2025-08-26 at 11 11 35 PM" src="https://github.com/user-attachments/assets/f3f7a968-5e2b-446b-b4fd-077801367cee" />

>  I'm not sure whether "Restore Default Styles…" should still be on its own on the left on Windows or if it should be the right-most option after "Add from File…"

I have a guess that it should be moved into the right group after the "Add from File" because we already have something similar in one of our dialogs on windows. If you use the "Change parent item" context menu option on an attachment, there are 3 buttons in the dialog, and they all appear together:

<img width="804" height="448" alt="Screenshot 2025-08-26 at 11 06 48 PM" src="https://github.com/user-attachments/assets/e981b66a-187b-4bff-b227-10e41b0cc17a" />

For comparison, this same dialog on macOS looks like this:

<img width="800" height="453" alt="Screenshot 2025-08-26 at 11 07 24 PM" src="https://github.com/user-attachments/assets/2a58213a-dfdb-46fc-9338-02c25d6dd317" />

This seems somewhat similar to our use case here, I think? So then this is what windows looks like:

<img width="592" height="397" alt="Screenshot 2025-08-26 at 11 10 44 PM" src="https://github.com/user-attachments/assets/9616253c-b11b-4baf-a263-46a1de1600a3" />

dstillman commented:
> we do have other instances of Backspace/Delete on table rows performing some kind of "Remove"-like action: in library tab, Backspace on a row removes items from collections (or trashed them), in citation dialog, backspace on a row will remove all bubbles of that item from the citation, and in the (in-works) redesigned edit bibliography dialog, backspace will remove the item from the bibliography. Not to mention that pressing Backspace vs Tab (to focus the - icon) + Space/Enter (to click it) is just a bit faster.

The problem is just the button. Backspace (or Ctrl-Backspace) is obviously a common action in a list, but none of those other examples (at least other than Edit Bibliography, which I haven't tested) have a (-) button that's triggered by Backspace. The new citation dialog has a (+), but that's the default action for the row that you get on double-click or Enter. It seems weird and confusing to me to have a *non-default* action that's triggered by a button that you can't access.

And now that we don't have a proper selected state, I think people will be even less likely to realize that Backspace works.

Deleting styles will also be much more rare than wanting to see (or hear) what styles are installed, so in most cases people will still just use the arrow keys.

So I think we need to put the buttons in the tab order. This shouldn't apply to other lists, since they're not doing the same thing.

(While we're at it, Page-Up/Down page the containing preferences pane here rather than the styles list as they should when the list has focus.)

dstillman commented:
That's better.

> When I press Page-Up/Down after tabbing into the table, it scrolls the list to the top/bottom as, I believe, it is supposed to

Are you pressing Page-Up/Page-Down or Home/End? Home/End (Fn-left-arrow and Fn-right-arrow) are working properly for me. Page-Up/Page-Down (Fn-up-arrow and Fn-down-arrow) aren't.

https://github.com/user-attachments/assets/e8283af1-7790-4826-bb14-0b97ed99bbb0

dstillman commented:
I've updated the intro text:

> Zotero can generate citations and bibliographies in over 10,000 citation styles. Add styles here to make them available when selecting styles throughout Zotero.

abaevbog commented on chrome/content/zotero/preferences/preferences_cite.jsx:
```diff
@@ -97,10 +114,18 @@ Zotero_Preferences.Cite = {
 			];
 			var handleKeyDown = (event) => {
-				if (event.key == 'Delete' || Zotero.isMac && event.key == 'Backspace') {
+				if (event.key == 'Delete' || event.key == 'Backspace') {
```
Gotcha, I reverted this change, and I'll keep this convention in mind for the future.

I also added https://github.com/zotero/zotero/pull/5526 to delete rows on Delete vs Backspace in the `itemTree` of the citation dialog.

abaevbog commented:
> Zotero can generate citations and bibliographies in over 10,000 citation styles. Add styles here to make them available when selecting styles throughout Zotero.

I gotta say, this sounds great!

> Page-Up/Page-Down (Fn-up-arrow and Fn-down-arrow) aren't.

Oh, I see now. It looks like `PageUp`/`PageDown` were intentionally excluded from the list of events that prevent scrolling on macOS in https://github.com/zotero/zotero/commit/e68685521b17926117b2ec9bf585b79011fa198a. 

https://github.com/zotero/zotero/blob/afc02141cedd7f86974a782811b49e9a2068abdd/chrome/content/zotero/components/virtualized-table.jsx#L341-L343

I'm not sure I understand why... From what I see, if  `PageUp` and `PageDown` were added on macOS to `preventScrollKeys`, everything would work as expected without scrolling the preferences pane. And it doesn't seem to affect selection in any unwanted manner as the commit description indicates. 

@adomasven, is there something with PageUp and PageDown handling on macOS I am missing that doesn't allow us to stop their propagation?

adomasven commented:
> @adomasven, is there something with PageUp and PageDown handling on macOS I am missing that doesn't allow us to stop their propagation?

Per @dstillman request https://github.com/zotero/zotero/pull/2029, which was the old XUL tree, and also the current Finder behavior on macOS - scrolling the view, instead of selecting a row one page up/down.

I guess the event bubbles up to the parent element, so preventing that should be enough?

abaevbog commented:
Gotcha, thank you for the explanation. Right, from what I see, the behavior of just scrolling the tree without changing the selection is enforced in `_onKeyDown`:

https://github.com/zotero/zotero/blob/afc02141cedd7f86974a782811b49e9a2068abdd/chrome/content/zotero/components/virtualized-table.jsx#L577-L584

While `preventScrollKeys` is just a set of keys whose events get `stopPropagation()` and `preventDefault()`. If I add `pageUp` and `pageDown` into that set on macOS, everything seems to still work fine in the itemTree, plus it would fix the issue of scrolling the table's container. I just pushed that little tweak. Does that seem reasonable?

https://github.com/user-attachments/assets/80e648ec-e485-4fa9-a8a0-ffbd1b7f85fe

abaevbog commented:
On a separate note, earlier in this PR, I removed the pointer-events workaround from virtualized table to make it easier to add handling for buttons in cells without JS managing every effect. I also cleaned up a few functions that seemed no longer necessary after that (e.g. the manual tooltip handling). It all happens in https://github.com/zotero/zotero/pull/5518/commits/af24c80f56c99b50e9883ae4f0c62eee1e338638 and https://github.com/zotero/zotero/pull/5518/commits/3461b932d9412d263b05b70216018cf70c0b6330. I've used the itemTree quite a bit with this change without any problems - but let me know if you see any issues with this.

abaevbog commented:
I see what you mean. The focus ring is really just a more subtle styling for the `.selected` row, so it appeared on just whichever row was selected.

I tweaked it to only show it when the table itself has focus. How does that look?

https://github.com/user-attachments/assets/d7b31066-dcf6-45ac-968f-100d6192198d

dstillman commented:
Merged
